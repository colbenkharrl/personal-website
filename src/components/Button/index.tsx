import Link from 'next/link';
import clsx from 'clsx';

const variantStyles = {
  primary:
    'bg-orange-500 font-semibold text-zinc-100 hover:bg-orange-600 active:bg-orange-500 active:text-zinc-100/70 dark:bg-orange-400 dark:hover:bg-orange-500 dark:active:bg-orange-700 dark:active:text-zinc-100/70',
  secondary:
    'bg-zinc-50 font-medium text-zinc-900 hover:bg-zinc-100 active:bg-zinc-100 active:text-zinc-900/60 dark:bg-zinc-800/50 dark:text-zinc-300 dark:hover:bg-zinc-800 dark:hover:text-zinc-50 dark:active:bg-zinc-800/50 dark:active:text-zinc-50/70',
  link: 'font-medium text-zinc-900 active:text-zinc-900/60 dark:text-zinc-300 dark:hover:text-zinc-50 dark:active:text-zinc-50/70',
};

export type ButtonProps = {
  /**
   * Style variant for button presentation.
   */
  variant?: keyof typeof variantStyles;
} & (
  | (React.ComponentPropsWithoutRef<'button'> & { href?: undefined })
  | React.ComponentPropsWithoutRef<typeof Link>
);

/**
 * Standard UI button with two themed style variants.
 */
export function Button({
  variant = 'primary',
  className,
  ...props
}: ButtonProps) {
  className = clsx(
    'disabled:opacity-30 inline-flex items-center gap-2 justify-center rounded-md py-2 px-3 text-sm outline-offset-2 transition active:transition-none',
    variantStyles[variant],
    className,
  );

  return typeof props.href === 'undefined' ? (
    <button className={className} {...props} />
  ) : (
    <Link className={className} {...props} />
  );
}
