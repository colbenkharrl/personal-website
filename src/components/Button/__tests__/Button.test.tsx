import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import * as stories from '../__stories__/Button.stories';
import { composeStories } from '@storybook/react';

const { Primary, Secondary, Link } = composeStories(stories);

it('Button matches snapshot (regression test)', () => {
  render(
    <div data-testid="Button">
      <Primary />
      <Secondary />
    </div>,
  );
  expect(screen.getByTestId('Button')).toMatchSnapshot(
    'Full component snapshot.',
  );
});

it('should render a <button> element if no href', () => {
  const { container } = render(<Primary />);

  expect(container.querySelector('button')).not.toBeNull();
  expect(container.querySelector('a')).toBeNull();
});

it('should render a <a> element if href present', () => {
  const { container } = render(<Link />);

  expect(container.querySelector('button')).toBeNull();
  expect(container.querySelector('a')).not.toBeNull();
});

it('should execute onClick prop when clicked', async () => {
  const spy = jest.fn();

  render(<Primary data-testid="Button" onClick={spy} />);

  fireEvent.click(screen.getByTestId('Button'));

  await waitFor(() => expect(spy).toHaveBeenCalledTimes(1));
});
