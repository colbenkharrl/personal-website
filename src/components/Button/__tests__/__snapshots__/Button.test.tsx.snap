// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot (regression test): Full component snapshot. 1`] = `
<div
  data-testid="Button"
>
  <script>
    !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&true)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}if(e==='light'||e==='dark')d.style.colorScheme=e}catch(e){}}()
  </script>
  <button
    class="disabled:opacity-30 inline-flex items-center gap-2 justify-center rounded-md py-2 px-3 text-sm outline-offset-2 transition active:transition-none bg-orange-500 font-semibold text-zinc-100 hover:bg-orange-600 active:bg-orange-500 active:text-zinc-100/70 dark:bg-orange-400 dark:hover:bg-orange-500 dark:active:bg-orange-700 dark:active:text-zinc-100/70"
  >
    Primary Button
  </button>
  <script>
    !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&true)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}if(e==='light'||e==='dark')d.style.colorScheme=e}catch(e){}}()
  </script>
  <button
    class="disabled:opacity-30 inline-flex items-center gap-2 justify-center rounded-md py-2 px-3 text-sm outline-offset-2 transition active:transition-none bg-zinc-50 font-medium text-zinc-900 hover:bg-zinc-100 active:bg-zinc-100 active:text-zinc-900/60 dark:bg-zinc-800/50 dark:text-zinc-300 dark:hover:bg-zinc-800 dark:hover:text-zinc-50 dark:active:bg-zinc-800/50 dark:active:text-zinc-50/70"
  >
    Secondary Button
  </button>
</div>
`;
